### ISSUE TITLE
Set up the surface on the compositor `new_surface` signal

### ISSUE THREAD

**Issue by acrisci** [swaywm/wlroots-rs#175]
Right now it seems like surfaces are only created (ie with `Surface::new`) when the surface comes from a shell. Not all surfaces will appear this way. For instance, the surfaces for client cursors.

If I call `SurfaceHandle::from_ptr` on a surface that has not previously been created with `Surface::new`, I get a crash.

Setting up the surface in the on the Compositor `new_surface` event makes much more sense for this because this is the time that the surface is actually created. Then it can be gotten from the pointer in other places safely.

I've run into this issue while implementing the `set_cursor` event of the seat.

**Comment by Timidger** [comment]
You get a crash because, in `Surface::new`, we set the `data` so that we can reconstruct the `Rc<Cell<bool>>` (which keeps track of how many references there are and if any of them have are being borrowed) from a plain `wlr_surface` pointer (which is how wlroots gives it to us).

If we do this instead in `new_surface` we will need a place for these `Surface`s to live, which is the only problem with that approach. We could then transition to a different way of storing surfaces, e.g. a `HashMap<*mut wlr_surface, Surface>` (so that when we get a surface pointer we either construct it if it's not found or we get a handle to the surface in the hash map).

Assuming I understand the requirements correctly, this seems like a fairly big change. It's possible to do however. The best place to store the hash map is probably in a hidden variable in `Compositor` since that is always accessible.

**Comment by acrisci** [comment]
I'm not necessarily against the hashmap idea, but I don't understand why we can't continue setting up the surface from the user data pointer as we do now.

The wlr_compositor will give us a `new_surface` event whenever a surface appears. We can call `Surface::new` there to set up the user data, and then replace all of the other `Surface::new` calls that appear in the code with `Surface::from_ptr`. That seems simpler to me. Will that work?

**Comment by Timidger** [comment]
[Surface has a drop impl](https://github.com/swaywm/wlroots-rs/blob/master/src/types/surface/surface.rs#L352). If we don't store the result of `new` somewhere it will be freed immediately.

If we `mem::forget` the Surface it will be accessible but then when the surface is destroyed we will not destroy the `Surface` properly, resulting in a dangling pointer that is accessible.

**Comment by acrisci** [comment]
Ah ok I see. So the model seems to assume that the surface role object is the sole owner of the surface and that the surface is no longer valid at the time the role object goes out of scope. This is practical if you assume everything is a shell surface, but client cursors and drag icons do not have role objects like that. They are no longer valid when the surface itself is destroyed which happens on the surface `destroy` event.

**Comment by Timidger** [comment]
There is an alternative, though it could be annoying.

We could return a `Surface` but wrap it in a structure that has a `PhantomData` that requires it to live only as long as the object it's accessed from (e.g. the cursor). It could look like this:

```rust
struct SurfaceWrapper<'surface> {
    surface: Surface,
    phantom: PhantomData<&'surface ()>
}

// So now we can do my_surface_wrapper.subsurfaces or any other function on the Surface
impl <'surface> Deref for SurfaceWrapper<'surface> {
    fn deref(&'surface self) -> &'surface Surface {
         &self.surface
    }
}
````

**Comment by acrisci** [comment]
I don't think we're going to be able to get around listening to surface `destroy` because the surface can literally be destroyed at any time, including when it is currently being used as a client cursor.

**Comment by acrisci** [comment]
Maybe we can just keep a hashmap of client cursors (and then drag icons) on the compositor, then remove it from the hashmap when the client cursor surface emits `destroy` which will cause it to drop and not be usable anymore.

**Comment by acrisci** [comment]
Yeah ok, that sounds interesting. Can you sketch that out for me so I can try to implement it?

**Comment by acrisci** [comment]
It just needs to go in here:

```
    pub fn set_surface<'this, O>(&'this mut self, surface: O, hotspot_x: i32, hotspot_y: i32)
        where O: Into<Option<&'this Surface>>
```

**Comment by Timidger** [comment]
Make it store a surface handle instead, and revert to an xcursor when you can't upgrade it?

**Comment by acrisci** [comment]
We'll need to listen to the `commit` signal as well for damage tracking for the drag icons. Popups (and I think subsurfaces on any shell) will need this as well. I need to look into subsurface damage tracking on commit, because that might necessitate having the surface be a standalone thing. Unless if the toplevel gets like a "child committed" event for that purpose.

I'm probably going to need to prioritize output damage tracking and drag icons to make sure it doesn't stretch the interface too much.

So I guess we have to choices:

1) Make the `Surface` more of a standalone struct that is not directly tied to the role objects. It will require its own handlers for events and need to be stored in a list on the compositor to keep references.
2) Create artificial role objects for things like client cursors and drag icons which manage the lifetime of the surface and take handlers.

**Comment by Timidger** [comment]
I like 1 more, since:
* No weird special cases for things that aren't really special cases
* Actually models how it works
* Can mean current role objects just utilize `SurfaceHandle` (by happily using `from_ptr` with the knowledge it's been setup properly) which is _usually_ what it wants to do because on the role objects you can only get a `SurfaceHandle` not a `&Surface`.

This is going to be a big change, but I think it's worth it. Thanks for bringing it to my attention @acrisci! 

RE: output damage tracking. I have a branch open that, while it hasn't been touched since around early May, has stubbed out quite a bit of that work.

Is this what you're going to work on next, or are you going to prioritize damage tracking and drag icons first? I can work on this if you can't.

**Comment by acrisci** [comment]
This needs to be done before work on either client cursors, drag icons, or output damage can be done. If you want to take it you can go ahead, otherwise I can work on it. If you can split it up so we can both work on it in parallel that would be good too.

**Comment by Timidger** [comment]
> This needs to be done before work on either client cursors, drag icons, or output damage can be done.

Gotcha.

> If you can split it up so we can both work on it in parallel that would be good too.

This is sort of hard to split up, since there is one "atomic" change here (moving the `Surface` ownership) but it affects a lot of things.

I'll plan to start it today after work. If this is a blocker for you though (and since you are in the future on the east coast) you can start it and post the branch here and I can take a look and help finish it later today.

**Comment by Timidger** [comment]
[Started working on it](https://github.com/swaywm/wlroots-rs/tree/surface-standalone)

**Comment by acrisci** [comment]
Oh ok. Can you make that a PR so it's easier to look at?

**Comment by Timidger** [comment]
Fixed with #184
